src/maintenance.js

```diff
@@ -0,0 +1,159 @@
+let taskcluster = require('taskcluster-client');
+let _ = require('lodash');
+let slugid = require('slugid');
+let assert = require('assert');
+let Debug = require('debug');
+
+let setPulseUser = async function({username, password, namespace, rabbitManager, cfg}) {
+  await rabbitManager.createUser(username, password, cfg.app.userTags);
+
+  await rabbitManager.setUserPermissions(
+    username,
+    cfg.app.virtualhost,
+    cfg.app.userConfigPermission.replace(/{{namespace}}/g, namespace),
+    cfg.app.userWritePermission.replace(/{{namespace}}/g, namespace),
+    cfg.app.userReadPermission.replace(/{{namespace}}/g, namespace),
+  );
+};
+
+/*
+ * Attempt to create a new namespace entry and associated Rabbit user.
+ * If the requested namespace exists, update it with any user-supplied settings,
+ * and return it.
+ */
+module.exports.claim = async function({Namespace, cfg, rabbitManager, namespace, contact, expires}) {
+  let newNamespace;
+  let created;
+
+  try {
+    newNamespace = await Namespace.create({
+      namespace: namespace,
+      username: namespace,
+      password: slugid.v4(),
+      created: new Date(),
+      expires,
+      rotationState:  '1',
+      nextRotation: taskcluster.fromNow(cfg.app.namespaceRotationInterval),
+      contact,
+    });
+
+    created = true;
+  } catch (err) {
+    if (err.code !== 'EntityAlreadyExists') {
+      throw err;
+    }
+
+    created = false;
+
+    // get the existing row
+    newNamespace = await Namespace.load({namespace: namespace});
+
+    // If this claim contains different information, update it accordingly
+    if (!_.isEqual(
+      {expires: newNamespace.expires, contact: newNamespace.contact},
+      {expires, contact})) {
+      await newNamespace.modify(entity => {
+        entity.expires = expires;
+        entity.contact = contact;
+      });
+
+      newNamespace = await Namespace.load({namespace: namespace});
+    }
+  }
+
+  if (created) {
+    // set up the first user as active,
+    await setPulseUser({
+      username: `${namespace}-1`,
+      password: newNamespace.password,
+      namespace, cfg, rabbitManager});
+    // ..and the second user as inactive (empty string means no logins allowed)
+    await setPulseUser({
+      username: `${namespace}-2`,
+      password: '',
+      namespace, cfg, rabbitManager});
+  }
+
+  return newNamespace;
+};
+
+module.exports.delete = async function({Namespace, rabbitManager, cfg, namespace}) {
+  let debug = Debug('delete');
+
+  // use the configuration regexp to determine if an object is owned by this user
+  let owned = new RegExp(cfg.app.userConfigPermission.replace(/{{namespace}}/g, namespace));
```

Comment:
How much guarding should we be doing on this namespace field? If it is accidentally an empty string or something will bad stuff happen?